configstack/stack_plan.go

```diff
@@ -0,0 +1,167 @@
+	fmt.Fprintf(terragruntOptions.Writer, "%s\nSummary:\n", separator)
+	for _, result := range results {
+		errMsg := ""
+		if result.Err != nil {
+			errMsg = fmt.Sprintf(", Error: %v", result.Err)
+		}
+
+		fmt.Fprintf(terragruntOptions.Writer, "    %v : %v%v\n", result.Module.Path, result.Message, errMsg)
+	}
+}
+
+// Check the output message
+func warnAboutMissingDependencies(module TerraformModule, output string) {
+	if strings.Contains(output, ": Resource 'data.terraform_remote_state.") {
+		var dependenciesMsg string
+		if len(module.Dependencies) > 0 {
+			dependenciesMsg = fmt.Sprintf(" contains dependencies to %v and", module.Config.Dependencies.Paths)
+		}
+		module.TerragruntOptions.Logger.Printf("%v%v refers to remote state "+
+			"you may have to apply your changes in the dependencies prior running terragrunt plan-all.\n",
+			module.Path,
+			dependenciesMsg,
+		)
+	}
+}
+
+// Parse the output message to extract a summary
+func extractSummaryResultFromPlan(output string) (string, int) {
+	const noChange = "No changes. Infrastructure is up-to-date."
+	if strings.Contains(output, noChange) {
+		return "No change", 0
+	}
+
+	result := planResultRegex.FindStringSubmatch(output)
+	if len(result) == 0 {
+		return "Unable to determine the plan status", -1
+	}
+
+	// Count the total number of changes
+	sum := 0
+	for _, value := range result[1:] {
+		count, _ := strconv.Atoi(value)
+		sum += count
+	}
+	if sum != 0 {
+		return result[0], sum
+	}
+
+	// Sometimes, terraform returns 0 add, 0 change and 0 destroy. We return a more explicit message
+	return "No effective change", 0
+}
+
+// This is used to return the total number of changes if -detail-exitcode is specified and return an exit code of 2
+// to be compliant with the terraform specification.
+type countError struct{ count int }
+
+func (err countError) Error() string {
+	article, plural := "is", ""
+	if err.count > 1 {
+		article, plural = "are", "s"
+	}
+	return fmt.Sprintf("There %s %v change%s to apply", article, err.count, plural)
+}
+
+// If there are changes, the exit status must be = 2
+func (err countError) ExitStatus() (int, error) {
+	if err.count > 0 {
+		return CHANGE_EXIT_CODE, nil
+	}
+	return NORMAL_EXIT_CODE, nil
+}
+
+// This is a specialized version of MultiError type
+// It handles the exit code differently from the base implementation
+type PlanMultiError struct {
+	MultiError
+}
+
+func (this PlanMultiError) ExitStatus() (int, error) {
+	exitCode := NORMAL_EXIT_CODE
+	for i := range this.Errors {
+		if code, err := shell.GetExitCode(this.Errors[i]); err != nil {
+			return UNDEFINED_EXIT_CODE, this
+		} else if code == ERROR_EXIT_CODE || code == CHANGE_EXIT_CODE && exitCode == NORMAL_EXIT_CODE {
+			// The exit code 1 is more significant that the exit code 2 because it represents an error
+			// while 2 represent a warning.
+			return UNDEFINED_EXIT_CODE, this
+		}
+	}
+	return exitCode, nil
+}
```

Comment:
You could have unit tests or integration tests check that `PlanMultiError` and `countError` handle exit codes as expected.